c++/src/Compression.cc

```diff
@@ -187,45 +161,81 @@ namespace orc {
 
     virtual bool Next(void** data, int* size) override;
     virtual std::string getName() const override = 0;
+    virtual void BackUp(int count) override;
+    virtual void suppress() override;
+    virtual uint64_t flush() override;
 
    protected:
     // return total compressed size
     virtual uint64_t doStreamingCompression() = 0;
+
+    // Buffer to hold uncompressed data until user calls Next()
+    BlockBuffer rawInputBuffer;
   };
 
+  void CompressionStream::BackUp(int count) {
+    uint64_t backup = static_cast<uint64_t>(count);
+    uint64_t currSize = rawInputBuffer.size();
+    if (backup > currSize) {
+      throw std::logic_error("Can't backup that much!");
+    }
+    rawInputBuffer.resize(currSize - backup);
+  }
+
+  uint64_t CompressionStream::flush() {
+    void* data;
+    int size;
+    if (!Next(&data, &size)) {
+      throw std::runtime_error("Failed to flush compression buffer.");
+    }
+    BufferedOutputStream::BackUp(outputSize - outputPosition);
+    rawInputBuffer.resize(0);
+    outputSize = outputPosition = 0;
+    return BufferedOutputStream::flush();
+  }
+
+  void CompressionStream::suppress() {
+    outputBuffer = nullptr;
+    outputPosition = outputSize = 0;
+    rawInputBuffer.resize(0);
+    BufferedOutputStream::suppress();
+  }
+
   CompressionStream::CompressionStream(OutputStream* outStream, int compressionLevel,
                                        uint64_t capacity, uint64_t blockSize, MemoryPool& pool,
                                        WriterMetrics* metrics)
-      : CompressionStreamBase(outStream, compressionLevel, capacity, blockSize, pool, metrics) {
+      : CompressionStreamBase(outStream, compressionLevel, capacity, blockSize, pool, metrics),
+        rawInputBuffer(pool, blockSize) {
     // PASS
   }
 
   bool CompressionStream::Next(void** data, int* size) {
-    if (bufferSize != 0) {
+    if (rawInputBuffer.size() != 0) {
```

Comment:
Maybe we can leverage the flexibility of BlockBuffer Here. That is, we can do compress only when the size of rawInputBuffer reachs compression block buffer size.

Reply:
Great advice, I'll address it in my next work project.